### PR TITLE
[#864] feat(server): Introduce Jersey to strengthen REST API

### PR DESCRIPTION
**Pull request by xianjingfeng**
### What changes were proposed in this pull request?
Introduce Jersey to strengthen REST API.
To use a new version of Jersey and being compatible with Hadoop 2.x, we have to use a shaded version of Jersery.
However Uniffle doesn't do binary release currently, we choose to introduce hbase's one as a temporary solution.

todo: release a shaded Jersey package for Uniffle's internal use or migrate to Graddle.

### Why are the changes needed?
Fix: #864 

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added UT.

### CODE REVIEW

**Comment by xianjingfeng** [comment]
1. Jersey1 is used in this PR. I have try to use jersey2, but i found it is not compatible with hadoop. 
2. We can't use lambda expression in jersey resources with jersey 1.9. https://stackoverflow.com/questions/24232784/jersey-and-java-8-lambda-expression

**Comment by jerqi** [comment]
SGTM, Let's merge this after #768 #973.

File: common/src/main/java/org/apache/uniffle/common/web/resource/PrometheusMetricResource.java
Comment (xianjingfeng) [reply common/src/main/java/org/apache/uniffle/common/web/resource/PrometheusMetricResource.java:78]: ok

File: common/pom.xml
Comment (xianjingfeng) [reply common/pom.xml:150]: For returning json
https://github.com/apache/incubator-uniffle/blob/598daf9a1f30bb254cf8bfbdb2040d4d01fc9674/coordinator/src/main/java/org/apache/uniffle/coordinator/web/resource/ServerResource.java#L79

**Comment by xianjingfeng** [comment]
> > We can't use lambda expression in jersey resources with jersey 1.9. https://stackoverflow.com/questions/24232784/jersey-and-java-8-lambda-expression
> 
> Is it possible to upgrade Jersey to at least 1.9+? Otherwise, new code may introduce lambda, and that would be crushed in runtime...

I will try it.

**Comment by xianjingfeng** [comment]
> > We can't use lambda expression in jersey resources with jersey 1.9. https://stackoverflow.com/questions/24232784/jersey-and-java-8-lambda-expression
> 
> Is it possible to upgrade Jersey to at least 1.9+? Otherwise, new code may introduce lambda, and that would be crushed in runtime...

I have tried to upgrade jersey to 1.19, but it is not compatible with hadoop.

**Comment by jerqi** [comment]
> > > We can't use lambda expression in jersey resources with jersey 1.9. https://stackoverflow.com/questions/24232784/jersey-and-java-8-lambda-expression
> > 
> > 
> > Is it possible to upgrade Jersey to at least 1.9+? Otherwise, new code may introduce lambda, and that would be crushed in runtime...
> 
> I have tried to upgrade jersey to 1.19, but it is not compatible with hadoop.

Could we shade jersey and use the shaded jersey?

**Comment by xianjingfeng** [comment]
> Could we shade jersey and use the shaded jersey?

Yes, but i feel a little trouble. Currently, it just not compatible with hadoop when running UTs. Should we create a new repository for third-part libraries? Just like https://github.com/apache/hbase-thirdparty. Do you have any other good suggestions? @advancedxy @zuston @jerqi

**Comment by jerqi** [comment]
Maybe we can use Hbase shaded jar first.  https://mvnrepository.com/artifact/org.apache.hbase.thirdparty/hbase-shaded-jersey/4.1.4 But it's a little weird. I can survey whether we can have better solution. also cc @kaijchen

**Comment by yl09099** [comment]
I have solved this conflict problem in #963. Of course, I don't know if we are talking about the same problem. I can access the API interface normally by using Jersey2 @jerqi @xianjingfeng

**Comment by xianjingfeng** [comment]
> I have solved this conflict problem in #963. Of course, I don't know if we are talking about the same problem. I can access the API interface normally by using Jersey2 @jerqi @xianjingfeng

Have you tried running UTs that containing hadoop mini cluster?

**Comment by kaijchen** [comment]
> Maybe we can use Hbase shaded jar first. https://mvnrepository.com/artifact/org.apache.hbase.thirdparty/hbase-shaded-jersey/4.1.4 But it's a little weird. I can survey whether we can have better solution.

SGTM.

**Comment by yl09099** [comment]
My PR is not yet complete, but I just use my way to test the ServletTest, is to solve the  exception: Java. Lang. ArrayIndexOutOfBoundsException: 52264 @xianjingfeng

**Comment by xianjingfeng** [comment]
> My PR is not yet complete, but I just use my way to test the ServletTest, is to solve the exception: Java. Lang. ArrayIndexOutOfBoundsException: 52264 @xianjingfeng

This execption is cause by using lambda expression in  jersey1. The solution is to use jersey1.19 or jersey2.  But both of them are not compatible with hadoop.

**Comment by jerqi** [comment]
@pan3793  Did the Kyuubi encounter similar issues? How to solve it? Could you give us some suggestion?

**Comment by pan3793** [comment]
No, Kyuubi itself does not heavily rely on Hadoop libraries, we use Hadoop shaded client(only available on Hadoop3), so there is no conflict.

You may want to learn what Spark did before 3.4, it uses Jersey 2.x and still works with Hadoop 2 clients.

**Comment by jerqi** [comment]
> No, Kyuubi itself does not heavily rely on Hadoop libraries, we use Hadoop shaded client(only available on Hadoop3), so there is no conflict.
> 
> You may want to learn what Spark did before 3.4, it uses Jersey 2.x and still works with Hadoop 2 clients.

It is caused by usage of `MiniDFSCluster`. Spark don't use `MiniDFSCluster`. Thanks.

**Comment by yl09099** [comment]
> > My PR is not yet complete, but I just use my way to test the ServletTest, is to solve the exception: Java. Lang. ArrayIndexOutOfBoundsException: 52264 @xianjingfeng
> 
> This execption is cause by using lambda expression in jersey1. The solution is to use jersey1.19 or jersey2. But both of them are not compatible with hadoop.

What is the incompatibility problem you are referring to? So far, I have no problem in my tests

**Comment by jerqi** [comment]
@xianjingfeng Could we add new module like 
```
third-party
```
We shaded the jersey in this module. Other modules can depend on the module `third-party`.

**Comment by xianjingfeng** [comment]
> What is the incompatibility problem you are referring to? So far, I have no problem in my tests

You can try to run the UT which extends org.apache.uniffle.storage.HadoopTestBase

**Comment by xianjingfeng** [comment]
> @xianjingfeng Could we add new module like
> 
> ```
> third-party
> ```
> 
> We shaded the jersey in this module. Other modules can depend on the module `third-party`.

SGTM. I will try it.

**Comment by yl09099** [comment]
> HadoopTestBase

org.apache.uniffle.test.ServletTest，The unit test class is extended org.apache.uniffle.storage.HadoopTestBase,I had no problem with the test

**Comment by xianjingfeng** [comment]
> > HadoopTestBase
> 
> org.apache.uniffle.test.ServletTest，The unit test class is extended org.apache.uniffle.storage.HadoopTestBase,I had no problem with the test

Maybe you need fix this problom first. https://github.com/apache/incubator-uniffle/actions/runs/5368148043/jobs/9739499525?pr=963

**Comment by yl09099** [comment]
> > > HadoopTestBase
> > 
> > 
> > org.apache.uniffle.test.ServletTest，The unit test class is extended org.apache.uniffle.storage.HadoopTestBase,I had no problem with the test
> 
> Maybe you need fix this problom first. https://github.com/apache/incubator-uniffle/actions/runs/5368148043/jobs/9739499525?pr=963

Okay, I'll fix it

**Comment by yl09099** [comment]
> > > > HadoopTestBase
> > > 
> > > 
> > > org.apache.uniffle.test.ServletTest，The unit test class is extended org.apache.uniffle.storage.HadoopTestBase,I had no problem with the test
> > 
> > 
> > Maybe you need fix this problom first. https://github.com/apache/incubator-uniffle/actions/runs/5368148043/jobs/9739499525?pr=963
> 
> Okay, I'll fix it

I was wrong, I only solved the ServletTest problem, but there were many other conflicts

**Comment by jerqi** [comment]
> I found it a bit troublesome, we need to execute `mvn install` first and then execute other commands。https://github.com/xianjingfeng/incubator-uniffle/actions/runs/5374375885. Should we go ahead in this way? @jerqi

You should add the module in the parent pom like other modules. https://github.com/apache/incubator-uniffle/blob/master/pom.xml

**Comment by xianjingfeng** [comment]
> You should add the module in the parent pom like other modules. https://github.com/apache/incubator-uniffle/blob/master/pom.xml

I can't import classes by using shaded package name if we add the module in the parent pom. It seems that this shaded package name is only generated after compiling.

**Comment by pan3793** [comment]
> We shaded the jersey in this module. Other modules can depend on the module third-party.

This would be a tough way, at least when using Maven as the building system, and IDEA as IDE.

It used to work well with Gradle on IDEA, and is adopted by some projects like Iceberg. But unfortunately, the recent change of IDEA [IDEA-321592](https://youtrack.jetbrains.com/issue/IDEA-321592) breaks it totally.

**Comment by pan3793** [comment]
> I can't import classes by using shaded package name if we add the module in the parent pom. It seems that this shaded package name is only generated after compiling.

@xianjingfeng This is a known issue [IDEA-93855](https://youtrack.jetbrains.com/issue/IDEA-93855/Please-add-support-for-maven-shade-plugin-renaming) which existing over a decade.

**Comment by xianjingfeng** [comment]
> > I can't import classes by using shaded package name if we add the module in the parent pom. It seems that this shaded package name is only generated after compiling.
> 
> @xianjingfeng This is a known issue [IDEA-93855](https://youtrack.jetbrains.com/issue/IDEA-93855/Please-add-support-for-maven-shade-plugin-renaming) which has been existing over a decade.

Get it. Thanks.

**Comment by jerqi** [comment]
> > You should add the module in the parent pom like other modules. https://github.com/apache/incubator-uniffle/blob/master/pom.xml
> 
> I can't import classes by using shaded package name if we add the module in the parent pom. It seems that this shaded package name is only generated after compiling.

You're right. The jar is shaded when phase is package.

**Comment by jerqi** [comment]
@xianjingfeng If we don't have better solution, we can use the dependency `hbase-thirdparty`. We should guarantee that we can compile and run uts with the command `mvn package`. Because the Apache member will use similar command to verify the release. If we can't run directly, we need to explain much.

**Comment by yl09099** [comment]
> 

I should have no impact, I just need to call the RESTful interface

**Comment by xianjingfeng** [comment]
Already changed to use `hbase-thirdparty`.  PTAL @jerqi  @advancedxy @yl09099

**Comment by yl09099** [comment]
> Already changed to use `hbase-thirdparty`. PTAL @jerqi @advancedxy @yl09099



> Already changed to use `hbase-thirdparty`. PTAL @jerqi @advancedxy @yl09099

I also tried to solve it, but I introduced the hbase-shaded-jackson-jaxrs-json-provider package, you can not write the JsonConverter conversion class, I have resolved the conflict in #960, you see my way is okay?

**Comment by xianjingfeng** [comment]
> > Already changed to use `hbase-thirdparty`. PTAL @jerqi @advancedxy @yl09099
> 
> > Already changed to use `hbase-thirdparty`. PTAL @jerqi @advancedxy @yl09099
> 
> I also tried to solve it, but I introduced the hbase-shaded-jackson-jaxrs-json-provider package, you can not write the JsonConverter conversion class, I have resolved the conflict in #960, you see my way is okay?

I tried to register resource in the same way as you before, but I found that I couldn’t get HttpServletRequest or HttpServletRequest, have you tested it and it’s ok?

**Comment by xianjingfeng** [comment]
> Hi @xianjingfeng and @slfan1989 this and #963 which one should be merged first?

It doesn't matter which one is merged first. But #963  should not contain the content that introduces jersey.

> And by the way, seems like these two PRs both have PR failrures.

The failrure of this pr because of flaky test.

**Comment by slfan1989** [comment]
> Hi @xianjingfeng and @slfan1989 this and #963 which one should be merged first? And by the way, seems like these two PRs both have PR failrures.

@advancedxy @xianjingfeng From my personal point of view, this pr should be merged first. #963 is still under discussion, and yl09099 is currently working on the documentation. It might take some time before we can merge it.

**Comment by yl09099** [comment]
> The failrure of this pr because of flaky test.

This PR should be merged first

File: common/src/main/java/org/apache/uniffle/common/web/JerseyAutoDiscoverable.java
Comment (xianjingfeng) [reply common/src/main/java/org/apache/uniffle/common/web/JerseyAutoDiscoverable.java:27]: I have no idea.

**Comment by xianjingfeng** [comment]
> @xianjingfeng This looks good to me now. I modified some content of the PR summary. Please amend/update accordingly if there are anything missing or left. Then I think we can squash and merge this PR.

SGTM. Thanks for your work

**Comment by xianjingfeng** [comment]
Do you have any other suggestions? @jerqi @yl09099

**Comment by slfan1989** [comment]
> Do you have any other suggestions? @jerqi @yl09099

LGTM +1.

**Review by slfan1989 — APPROVED** [review]

**Comment by yl09099** [comment]
> Do you have any other suggestions? @jerqi @yl09099

LGTM +1.

**Comment by xianjingfeng** [comment]
Merged. Thanks all.

**Pull request merged**
